vta/tests/hardware/metal_test/device.py

```diff
@@ -0,0 +1,278 @@
+				elif insn[0] == "ALU":
+					uop_size = 1 if uop_compression else out_channels // VTA_BLOCK_OUT
+				self.uop_size_buf.append(uop_size)
+				self.action_buf.append(insn[0])
+		
+		# assert dimensions is divisible
+		assert batch % VTA_BATCH == 0
+		assert in_channels % VTA_BLOCK_IN == 0
+		assert out_channels % VTA_BLOCK_OUT == 0
+
+		# generate instructions
+		self.__generate_insn()
+
+		# uop setup
+		self.__uop_setup()
+
+	"""print the instructions for this device"""
+	def print_insn(self):
+		print_instruction(self.ins_size, self.insn_buf)
+	"""print the micro-ops for this device"""
+	def print_uop(self):
+		print_uop(np.sum(self.uop_size_buf), self.uop_buf)
+	"""print the parameters for this device"""
+	def print_params(self):
+		print_params()
+
+	def __generate_insn(self):
+		# previous state, present state and next state
+		prevs, ps, ns = "load", "load", "load"
+		self.insn_buf = (VTAGenericInsn * self.ins_size)()
+		index, cmp_ind = 0, 0
+		assert "LOAD" in self.insn_stream.insn_buf[0][0], "Instruction shold start with Load instruction"
+		SIZE = {"INP":self.inp_size, "WGT":self.wgt_size, "ACC":self.out_size, "OUT":self.out_size}
+
+		# generate instructions
+		for i in range(self.ins_size):
+			# check for order
+			if (i + 1) < self.ins_size: 
+				n_action, n_item = self.insn_stream.insn_buf[i+1][0], self.insn_stream.insn_buf[i+1][1]
+				# check for order
+				if "LOAD" in n_action:
+					ns = "load"
+				elif n_action == "GEMM" or n_action == "ALU":
+					ns = "compute"
+				elif "STORE" in n_action:
+					ns = "store"
+				elif n_action == "FINISH":
+					ns = "finish"
+
+				assert not(ps == "load" and ns == "store"), "Instruction Wrong order!"
+				assert not(ps == "compute" and ns == "load"), "Instruction Wrong order!"
+				assert not(ps == "store" and ns == "compute"), "Instruction Wrong order!"
+
+			# generate instructions for GEMM 
+			action, item, sram, dram, use_imm, imm_val, y_size, x_size, \
+			 x_stride, x_pad, y_pad, vec_len = self.insn_stream.insn_buf[i]
+			# size
+			size = SIZE.get(item) if (item != "UOP") else self.uop_size_buf[cmp_ind]
+
+			# dependencies
+			pop_next = 1 if (prevs == "store" and ps == "load") else 0
+			push_next = 1 if (ps == "load" and ns == "compute") else 0
+
+			# make sure dependancies, ACC don't need push_next dependencies
+			if action == "1DLOAD":
+				self.insn_buf[index] = get_1Dloadstore_insn(ACTIONS.get(action), ITEMS.get(item), sram, dram, size, 
+				 0, pop_next, 0, push_next)
+
+			elif action == "2DLOAD":
+				self.insn_buf[index] = get_2Dloadstore_insn(ACTIONS.get(action), ITEMS.get(item), sram, dram, 
+				 y_size, x_size, x_stride, y_pad, x_pad, 0, pop_next, 0, 0)
+
+			elif action == "GEMM":
+				self.insn_buf[index] = get_gemm_insn(0, (int) (self.batch / VTA_BATCH), (int) (self.in_channels / VTA_BLOCK_IN),
+				 (int) (self.out_channels / VTA_BLOCK_OUT), self.uop_compression, 1, 0, 0, 1)
```

Comment:
Do we want to program the push/pop dependences flags explicitly? Or add a function to the instruction stream class to add serialization flags automatically before the programming VTA?

Reply:
I'm not sure here, I'm assuming all the instructions should be in the order of load/compute store, therefore, the place where GEMM and ALU could be is pretty much set (dependencies are pretty much set?). I've checked the order of instructions before hand to see if the user instruction follows the correct pattern to prevent unwanted things to happen.